Inspect return value

isachit opened the pull request:
Added inspect_return_value.py and inspect_byproducts.

Issue Referenced - #1

isachit committed:
Add a README

isachit committed:
Update inspect_byproducts.py

isachit committed:
added comments

isachit committed:
added comments

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+
+  Inspections constitute an important part of in-toto. In this script,
+  we inspect the byproducts(such as stdout/stderr) of a step. The user supplies
+  an input string to the program(whose presence the user wants to test),
```
Can you try to rephrase this? Basically, what happens is that the script takes the path of a link file, a random string and an *operator*, which is used to compare a certain property of the link file with the random string.

isachit replied:
Fixed.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+
+  Suppose the link file is located at /user/abc/def/package.45gh325.link
+  and the user wants to check whether for the corresponding step(hence the link file),
+  the corresponding stderr field contains the string "test".
```
Minor comment: Please add a whitespace before parentheses.

isachit replied:
Fixed.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+         whether to check stdout or stderr field
+
+     presence:
+         is | is not | contains | contains not
```
I think `presence` is not an ideal name. Do you mind refactoring it to something like *operator* (above and below)?

isachit replied:
Fixed. Renamed it to operator.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+
+    <Purpose>
+    A function which performs the inspection as described above depending on various arguments.
+    Prints the boolean True or False depending upon the inspection result.
```
It is better to let this function return a boolean value instead of printing the strings "True" or "False.

isachit replied:
returns an Integer value now.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+import os
+import sys
+import argparse
+import inspect
```
What do you use `inspect` for?

isachit replied:
Fixed. Wanted to do something, but then forgot to remove.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+    args = parser.parse_args()
+
+    if args.verbose:
+      in_toto.log.logging.getLogger.setLevel(log.logging.INFO)
```
You don't really need the logger or the verbose flag for now. Because you don't have any log statements.

isachit replied:
I see. Will fix it in the next commit. Thank you!

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+
+    parser.usage = ("\n"
+                    "%(prog)s --link <path to link metadata>\n{0}"
+                    "[--stdout | --stderr]\n{0}"
```
This is not accurate. What you are actually doing is `--st <any string>`. Also note, that you never check the argument passed to `--st`. You just expect it to be  either "stdout" or "stderr", but if the user passes something else, you'll get a `KeyError`.

isachit replied:
Fixed. Now checking the arguments. If anything else than "stdout" or "stderr" is given, it returns integer value 2, and prints "Please specify only one from the following- stdout, stderr"

isachit replied:
2) KeyError - fixed. It raises KeyError when the corresponding field is empty.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+    parser.usage = ("\n"
+                    "%(prog)s --link <path to link metadata>\n{0}"
+                    "[--stdout | --stderr]\n{0}"
+                    "[--is | --is-not | --contains | --contains-not]\n{0}"
```
Same as above. Your argparser actually does `-p <any string>`.

isachit replied:
Fixed.

lukpueh replied:
The usage message still does not describe what the argparser actually parses.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+                    "%(prog)s --link <path to link metadata>\n{0}"
+                    "[--stdout | --stderr]\n{0}"
+                    "[--is | --is-not | --contains | --contains-not]\n{0}"
+                    "<string>\n{0}"
```
Your argparser actually does not treat `<string>` as a positional argument but as a keyword argument `-s <string>`.

isachit replied:
Fixed. Now treating it as a positional argument.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+
+
+    """
+    l = link_import.read_from_file(link)
```
You should exit gracefully if the the link file does not exist.

isachit replied:
Fixed. Returns integer value 2, and prints "The path specified to the link file is invalid"

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+
+    """
+    l = link_import.read_from_file(link)
+    std_out_err = l.byproducts[std]
```
What happens if  the link's `byproducts` dictionary does not have a key `std`? Note that you don't sanitize your input, so `std` could be anything. And if `std` were one of "stderr" or "stdout", the byproducts field could be empty.
Please always try to program defensively.

isachit replied:
Fixed.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+            print('False')
+      else:
+            print('True')
+
```
As mentioned above, your function should rather return the bool instead of printing it.

lukpueh replied:
Also, you should sanitize your `presence` input. You could allow different cases (e.g. "CONTAINS" instead of "contains"). And you should exit gracefully but verbosely if the input is wrong.

isachit replied:
It returns integer 0 - when true, 1 - when false, 2 - when there is some error in user supplied input as discussed above. Shall i change it to boolean?

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+
+    elif presence == 'contains':
+      if (std_out_err.find(inputstring) != -1):
+            print('True')
```
Please use consistent (2 spaces per indentation level, 4 for line continuation) indentations.

isachit replied:
Sorry. My bad. Will do so from now onwards.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,140 @@
+    if args.verbose:
+      in_toto.log.logging.getLogger.setLevel(log.logging.INFO)
+
+    inspect_byproducts(args.link, args.outerr, args.presence, args.string)
```
Since this is a command line script, it would be great the exit with a descriptive return value. I'd suggest `0` if the function returns `True`, `1` if it returns `False` and `2` (like argparse) if the user specified wrong arguments, or the link file could not be found.

isachit replied:
Fixed.

lukpueh reviewed (changes requested):
Thanks for your PR Sachit. It looks pretty much like what I wanted. Most of my comments are pretty minor. Please go ahead and address them first (see if some of them apply for your other script to), and then I'll continue reviewing.

Also please consider some overall  comments:
- [Wrap lines at 79 characters](https://github.com/secure-systems-lab/code-style-guidelines#maximum-line-length);
- mention somewhere that your code requires `in-toto` installed on the path, either in a `requirements.txt` (c.f. [in-toto web tool](https://github.com/in-toto/layout-web-tool/blob/wizard/requirements.txt#L2)), or at least somewhere in the document header; 
- and for future reference, try to make an effort on things like
    - PR title -- you call it "inspect return value", which is only part of the PR,
    - PR description -- a little more text does not hurt, also link to the issue that your PR addresses
    - commits and commit messages -- try to break down your commits into logical, describable chunks. for instance your commit "Add a README" (32259fa) does not only not add a README but actually adds pretty much the whole implementation...

isachit committed:
Bug fixes, changed the return type, added exceptions, and sanitized user inputs

isachit commented:
Thank you so much for reviewing extensively. I will improve on all the things mentioned.

SantiagoTorres commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,176 @@
+  See LICENSE for licensing information.
+
+<Requires>
+  in-toto - installed on the path
```
I'm not sure if we need this "installed on the path" here

isachit replied:
Removed.

SantiagoTorres commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,176 @@
+        1 - if the inspection is unsuccessful
+        2 - if the user supplied arguments are invalid
+
+
```
I don't think we need all of this whitespace.

SantiagoTorres commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,176 @@
+        return 2
+
+    else:
+      l = link_import.read_from_file(link)
```
I suggest we use a more descriptive variable name.

isachit replied:
link_imported ? I suggest we should use this one.

SantiagoTorres replied:
sound reasonable, maybe `imported_link` or so?

isachit replied:
Okay! On it.

SantiagoTorres commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,176 @@
+        raise KeyError("The field corresponding to " + std + " in the link "
+                                                             "file is empty.")
+
+      switch_dict = {'is': (std_out_err == inputstring),
```
I"m not sure if it's worth running all these if we only want "is", for example

isachit replied:
Alright. So i changed it from dictionary to if conditions. No dictionary required now. Thanks! The changes will be reflected in the next commit.

SantiagoTorres replied:
Cool! thanks 💯

SantiagoTorres commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,176 @@
+
+    args = parser.parse_args()
+    args.operator = args.operator.lower()
+    args.outerr = args.outerr.lower()
```
I wonder if we'd want to separate the argparse setup to a private function for readability

isachit replied:
@lukpueh shall I make a separate function for the argparse setup? What are your thoughts on the same?  I think @SantiagoTorres is right.

lukpueh replied:
I don't have a strong opinion on this. :) In in-toto I used the `main()` function for `argparse` setup because argparsing is only relevant when the program is run on the command line, and `main` is usually associated with the command line, c.f.
```python
if __name__ == "__main__":
  main()
```

lukpueh replied:
But Santiago is the boss. :P

SantiagoTorres replied:
I don't understand why not call something akin to `parse_args()` inside `main()`. Just for the sake of readability...

lukpueh replied:
As I said, nothing against it, the only reason why I haven't done it so far is that main() pretty much does not do anything else but argparsing. But let's not waste time on this. Please sachit go ahead and move the argparsing to an extra function.

isachit replied:
Alright. On it.

isachit replied:
Done!

SantiagoTorres commented on inspect_return_value.py:
```diff
@@ -0,0 +1,134 @@
+    --ge ... greater than or equal
+
+  General usage:
+  python inspect_return_value.py -l <path/to/the/link/file> -bsbo "<bash/style/boolean/operator>" -i <integer>
```
I'm not sure if we'd like to have `-bsbo`. It feels a little odd.

isachit replied:
will change it to -operator .

lukpueh replied:
Btw., I think it is a common practice to have a shorthand option with a single hyphen and a long one with two hyphens, e.g. `-o, --operator`

isachit replied:
Ah yes! Sorry.

SantiagoTorres commented on inspect_return_value.py:
```diff
@@ -0,0 +1,134 @@
+
+
+    l = link_import.read_from_file(link)
+    switch_dict = {'eq': (integer == l.return_value),
```
Same comment as above

SantiagoTorres reviewed (commented):
Thanks for your first contribution! I think this is going in a good direction. I hope you find my comments helpful 👍

isachit committed:
Fixed identation issues, used if-else instead of dictionary implementation, changed some parts of documentation, some variable names, and changed the return value to 3, whenever IOError occurs

isachit committed:
Fixed issues in inspect_return_value.py, indentation, documentation, changed the return type of the function, added sanitization checks on user input, will move argparse to a different function, in the next commit

isachit committed:
fixed documentation

isachit committed:
Fixed Documentationissues

isachit committed:
Fixed Documentation issues

isachit committed:
Fixed Documentation Issues

isachit committed:
Made a separate function for parsing the arguments

isachit committed:
Fixed Documentation issues, Added a separate function for parsing arguments in inspect_return_value.py

lukpueh commented on inspect_byproducts.py:
```diff
@@ -68,54 +65,58 @@ def inspect_byproducts(link, std, operator, inputstring):
-        print("The path to the link file is invalid")
-        return 2
+        print("The path to the link file is invalid.")
+        return 3
```
Beware of inconsistent indentation

lukpueh commented on inspect_byproducts.py:
```diff
@@ -68,54 +65,58 @@ def inspect_byproducts(link, std, operator, inputstring):
 
-    else:
-      l = link_import.read_from_file(link)
+    imported_link = link_import.read_from_file(link)
```
This could still raise an error, above you only check if the path exists. Even if it returns `True` you might not have to necessary access privileges, or it is not a valid link file (could be a directory or any other file).
I think the best way is to do exception handling, either here and/or where you call `inspect_byproducts`.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -68,54 +65,58 @@ def inspect_byproducts(link, std, operator, inputstring):
+      0 - True
+      1 - False
+      2 - Wrong input
+      3 - IOError (Link file is non existent/ path to the link file is invalid
```
Sorry for being nitpicky here, but I really think that this function should return the boolean `True` or `False` and raise an exception if an error occurs.

isachit replied:
Okay! Fixed!

lukpueh commented on inspect_byproducts.py:
```diff
@@ -68,54 +65,58 @@ def inspect_byproducts(link, std, operator, inputstring):
-          return 0
+    elif operator == 'contains not':
+      if std_out_err.find(inputstring) == -1:
+        return 0
```
You should add an `else` clause that raises an exception, if the operator is not one of above.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -68,54 +65,58 @@ def inspect_byproducts(link, std, operator, inputstring):
+      std_out_err = imported_link.byproducts[std]
+    except KeyError:
+      raise KeyError("The field corresponding to " + std + " in the link"
+                        " file is empty.")
```
Choose either double indentation (preferred) or aligned indentation for line continuation

lukpueh commented on inspect_byproducts.py:
```diff
@@ -133,29 +134,40 @@ def main():
                               type=str, required=True, help="whether the "
-                                  "stdout or stderr is, is not, \ "
-                                  "contains, contains not, the input string")
+                                                            "stdout or stderr is, is not, \ "
```
Please remove the backslash `\`

lukpueh commented on inspect_byproducts.py:
```diff
@@ -133,29 +134,40 @@ def main():
-                                  "stdout or stderr is, is not, \ "
-                                  "contains, contains not, the input string")
+                                                            "stdout or stderr is, is not, \ "
+                                                            "contains, contains not, the input string")
```
Please consider the maximum line length.=

lukpueh commented on inspect_byproducts.py:
```diff
@@ -133,29 +134,40 @@ def main():
+    args = parse_args()
+
     if (args.operator != 'is') & (args.operator != 'is not') \
           & (args.operator != 'contains') & (args.operator != 'contains not'):
```
Beware of the difference between bitwise and logical operators.

lukpueh replied:
Also, we use double indentation for line continuation (not tripple).

lukpueh commented on inspect_byproducts.py:
```diff
@@ -172,5 +184,6 @@ def main():
       inspect_byproducts(args.link, args.outerr, args.operator, args.string)
 
 
+
```
Returning an `int` from `main()` won't be visible on the command line. Use `sys.exit(<return code>)` instead, or if you want to be really cool, you could use `ArgumentParser.error(<message>)` in case a wrong argument was passed, which also prints the usage message.

As for `inspect_byproducts`, i would do something like:
```python
try:
  if inspect_byproducts(args.link, args.outerr, args.operator, args.string):
    # Everything went fine
    sys.exit(0) 
  else:
    # The inspection failed
    sys.exit(1) 

# We could even be more granular and exit with different codes for `KeyError`, `IOError` and other Exceptions
except Exception as e:
  sys.exit(<some value greater than 2 and lower than 127>)
```

isachit replied:
Alright, I did it using `sys.exit(<return code>)`. Shall I go for `ArgumentParser.error(<message>)` ?

isachit committed:
Changed the return type of inspect_byproduct function to boolean, used sys.exit(), exceptions added

isachit committed:
Changed the return type to boolean, Added exceptions

isachit committed:
Minor changes in documentation

SantiagoTorres commented on inspect_return_value.py:
```diff
@@ -0,0 +1,180 @@
+                              help="Link metadata file to use for inspection "
+                              "of the step")
+
+    in_toto_args.add_argument("-o", "--operator",
```
I'm assuming you'd like to have something like [this](https://docs.python.org/2/library/argparse.html?highlight=argparse#choices) to limit the choices that can be passed.

isachit replied:
Yeess!! Perfect! Thank you so much!

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,180 @@
+      if std_out_err.find(inputstring) == -1:
+        return True
+    else:
+      raise Exception('Invalid operator')
```
It's helpful to add what was passed and what it should be to the exception message.

isachit replied:
I have used the argparser's `choice`. So it prints the operator supplied, and suggests the valid options to choose from.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,180 @@
+import securesystemslib.exceptions
+
+
+def inspect_byproducts(link, std, operator, inputstring):
```
Consider renaming `std` to `type` (type of byproduct, i.e. stdout or stderr), and changing `inputstring` to `input_string`.

isachit replied:
Done.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,180 @@
+                              help="Link metadata file to use for inspection "
+                              "of the step")
+
+    in_toto_args.add_argument("-st", "--outerr",
```
Maybe you could also change this to `-t`, `--type`, as in type of byproduct?

isachit replied:
Changed!

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,180 @@
+
+    else:
+      if args.verbose:
+        in_toto.log.logging.getLogger.setLevel(log.logging.INFO)
```
I think I argued that we don't really need the verbose flag, right?

isachit replied:
Removed.

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,180 @@
+
+    if (args.outerr != 'stdout') & (args.outerr != 'stderr'):
+      print('Please specify only one of the following - stdout, stderr')
+      sys.exit(3)
```
You are still using bitwise `&`, but you should use logical `&&`. [Argparser's choice argument](https://github.com/in-toto/in-toto-inspections/pull/2#discussion_r121991634) would eliminate the need for this check.

isachit replied:
Yes. It eliminates the need for the check.

lukpueh reviewed (changes requested):
A couple of more minor comments. Please be careful to address all comments. You can always ask if something is not clear. Also see if a comment might apply to other parts of the PR.

isachit committed:
Used arparser's choice method, changed the symbol's used for type, and other minor changes

isachit commented:
@lukpueh for the usage message. I have changed it. You can check and let me know if I correctly understood you. I think you meant adding some information like `--type of field to be inspected (--stdout | --stderr)` Did I get it right?

SantiagoTorres commented on inspect_return_value.py:
```diff
@@ -0,0 +1,170 @@
+    inspect_return_value to inspect the return value.
+
+    """
+    args = parse_args()
```
nice 👍

lukpueh commented on inspect_return_value.py:
```diff
@@ -0,0 +1,170 @@
+
+  Suppose the link file is located at /user/abc/def/package.45gh325.link
+  and the user wants to check whether for the corresponding step(hence the link file),
+  the corresponding return field contains an integer less than 5.
```
[Ping](https://github.com/in-toto/in-toto-inspections/pull/2#discussion_r120908370)

lukpueh commented on inspect_return_value.py:
```diff
@@ -0,0 +1,170 @@
+    elif operator == 'ge':
+      if integer <= imported_link.return_value:
+        return True
+
```
[Ping](https://github.com/in-toto/in-toto-inspections/pull/2#discussion_r121199600)

lukpueh replied:
Comment also applies to `inspect_byproducts.py`

isachit replied:
I think, since we are using the choices in argparse, it won't reach to a stage that some other operator other than the available choices is passed to this function. Since, on doing that, while parsing, argparse will itself raise an exception. So this stage when an operator other than the ones in the choices list is entered, won't ever be reached. What do you think?

isachit replied:
@SantiagoTorres  here. I think this is what you meant, right?

lukpueh commented on inspect_return_value.py:
```diff
@@ -0,0 +1,170 @@
+    imported_link = link_import.read_from_file(link)
+    if operator == 'eq':
+      if integer == imported_link.return_value:
+        return True
```
Actually you could do `return integer == imported_link.return_value` (also below). But I think that's a matter of taste.

lukpueh commented on inspect_return_value.py:
```diff
@@ -0,0 +1,170 @@
+
+    parser.usage = ("\n"
+                    "%(prog)s --link <path to link metadata>\n{0}"
+                    "[--<operator>]\n{0}"
```
Change to `--operator ('eq' | 'ne' | 'lt' | 'le' | 'gt' | 'ge')`

lukpueh commented on inspect_return_value.py:
```diff
@@ -0,0 +1,170 @@
+
+    in_toto_args.add_argument("-o", "--operator", choices=['eq', 'ne',
+                              'lt', 'le', 'gt', 'ge'], type=str,
+                              required=True, help="The boolen operator "
```
Actually, only AND, OR, NOT, and AND NOT are boolean operators (I know I called them so in the issue). Maybe we should rename them to logical operators.

lukpueh commented on inspect_return_value.py:
```diff
@@ -0,0 +1,170 @@
+      else:
+        sys.exit(1)
+    except Exception as e:
+      print('The following error occured',e)
```
Please add a whitespace: `print('The following error occured', e)`

lukpueh replied:
Comment also applies to `inspect_byproducts.py`

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,163 @@
+
+    parser.usage = ("\n"
+                    "%(prog)s --link <path to link metadata>\n{0}"
+                    "[--type of field to be inspected (--stdout | --stderr)]\n{0}"
```
Change to:
`--type (stdout | stderr)`

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,163 @@
+    parser.usage = ("\n"
+                    "%(prog)s --link <path to link metadata>\n{0}"
+                    "[--type of field to be inspected (--stdout | --stderr)]\n{0}"
+                    "[--operator (--is | --is not | --contains | --"
```
I think it is better to not have whitespaces in the operator "words", because they'd have to be escaped. Maybe you can change to `is-not` and `contains-not`.

Also change the usage message to:
`--operator (is | is-not | contains | contains-not)`

Alternatively, as mentioned before you could make a [mutually exclusive option group](https://docs.python.org/2.7/howto/argparse.html#conflicting-options) for the operators.

lukpueh reviewed (changes requested):
Please respect the line length of 79 characters (hint: most editors let you display a vertical ruler)

isachit committed:
Changed the operators in inspect_byproducts.py to is, is-not, contains, contains-not. Wrapped lines at 79 characters.
Minor bug fixes in documentation

SantiagoTorres commented on inspect_byproducts.py:
```diff
@@ -110,24 +112,24 @@ def parse_args():
-    in_toto_args.add_argument("-o", "--operator", choices=['is', 'is not',
-                              'contains', 'contains not'], type=str,
+    in_toto_args.add_argument("-o", "--operator", choices=['is', 'is-not',
+                              'contains', 'contains-not'], type=str,
```
wouldn't this be `%(prog)s --link whatever.link --operator ( is | is-not | contains | contains-not)` or so?

isachit committed:
Added an else clause in both the scripts, which raises an exception, when an invalid operator is supplied.
This is important since the functions inspect_return_value and inspect_byproducts can also be used as an import without importing the whole CLI script.
So in those cases, when just the respective functions are used, it raises an Exception on supplying invalid operator

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,168 @@
+                    "%(prog)s --link <path to link metadata>\n{0}"
+                    "[--type (--stdout | --stderr)]\n{0}"
+                    "[--operator (--is | --is-not | --contains | --"
+                    "contains-not)]\n{0}"
```
This still isn't right. Please try you command line tool and adjust the usage message here and above accordingly.

isachit replied:
Alright. On it.

SantiagoTorres replied:
Another possibility is that you can just drop setting `usage` and have argparse format it for you:
```
(toto-python2) [santiago@LykOS in-toto-inspections]$ python inspect_byproducts.py                                      
usage: inspect_byproducts.py [-h] -l <Path to link metadata> -t                                                        
                             {stdout,stderr} -o            
                             {is,is-not,contains,contains-not}                                                         
                             <String to compare>   
```
To do this, I just did the following:

```diff 

--- a/inspect_byproducts.py
+++ b/inspect_byproducts.py
@@ -114,18 +114,19 @@ def parse_args():
 
     lpad = (len(parser.prog) + 1) * " "
 
-    parser.usage = ("\n"
-                    "%(prog)s --link <path to link metadata>\n{0}"
-                    "[--type (--stdout | --stderr)]\n{0}"
-                    "[--operator (--is | --is-not | --contains | --"
-                    "contains-not)]\n{0}"
-                    "string\n\n"
-                    .format(lpad))
+#   parser.usage = ("\n"
+#                   "%(prog)s --link <path to link metadata>\n{0}"
+#                   "[--type (--stdout | --stderr)]\n{0}"
+#                   "[--operator (--is | --is-not | --contains | --"
+#                   "contains-not)]\n{0}"
+#                   "string\n\n"
+#                   .format(lpad))
 
     in_toto_args = parser.add_argument_group("in-toto-inspection options")
 
     in_toto_args.add_argument("-l", "--link", type=str, required=True,
-                              help="Path to the link file to be inspected")
+                              help="Path to the link file to be inspected",
+                              metavar="<Path to link metadata>")
 
     in_toto_args.add_argument("-t", "--type", choices=['stdout', 'stderr'],
                               type=str, required=True, help="Type of "
@@ -139,7 +140,8 @@ def parse_args():
 
     in_toto_args.add_argument("string", type=str,
                               help="The string to compare with the specified "
-                              "byproduct in the specified link file")
+                              "byproduct in the specified link file",
+                              metavar="<String to compare>")
```

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,168 @@
+
+    in_toto_args.add_argument("-l", "--link", type=str, required=True,
+                              help="Link metadata file to use for inspection"
+                              " of the step")
```
I'd suggest: "Path to link file to be inspected"

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,168 @@
+
+    in_toto_args.add_argument("-t", "--type", choices=['stdout', 'stderr'],
+                              type=str, required=True, help="Whether stdout or"
+                              " stderr is a byproduct")
```
I'd suggest: "Type of byproduct to inspect (stdout or stderr)"

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,168 @@
+
+    in_toto_args.add_argument("string", type=str,
+                              help="The string to which the return value "
+                              "should be compared")
```
I'd suggest: "The string to compare with the specified byproduct in the specified link file"

lukpueh commented on inspect_return_value.py:
```diff
@@ -0,0 +1,172 @@
+    --lt ... less than
+    --le ... less than or equal
+    --gt ... greater than
+    --ge ... greater than or equal
```
Please make sure that your documentation (docstrings, usage, ...) is consistent with how your tool actually works:
`--eq` != `--operator eq` == `--<operator>`

isachit replied:
Oh man!! Shit. Sorry! Changed

lukpueh commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,168 @@
+        return True
+    else:
+      raise('Invalid Operator ' + operator + '. Valid operators: is | is-not |'
+        ' contains | contains-not')
```
This will raise an exception. But not the one you think. Try and see what happens if you hit this `else`.

lukpueh replied:
Same applies to `else` clause in `inspect_return_value`

lukpueh commented on inspect_return_value.py:
```diff
@@ -0,0 +1,172 @@
+
+    in_toto_args.add_argument("-l", "--link", type=str, required=True,
+                              help="Link metadata file to use for inspection "
+                              "of the step")
```
I'd suggest: "Path to link file to be inspected"

lukpueh reviewed (changes requested):
Thanks for the changes. Please take your time and carefully read through both of your scripts and see if any of the comments I have made so far (also in previous reviews) might still be unresolved.
Some of the comments don't only apply to the line directly above them.

isachit committed:
Edited the exceptions, changed descriptions of the arguments

isachit committed:
Changed the documentation

SantiagoTorres commented on inspect_return_value.py:
```diff
@@ -0,0 +1,172 @@
+      return integer <= imported_link.return_value
+
+    else:
+        raise Exception(
```
Are you sure this exception will hit? if there are options already set on the argparse module. I strongly suggest you test these out locally before pushing

isachit replied:
Hey santiago! Well, this exception will not hit, if we use the tool. But in case, somebody imports just the function ```inspect_return_value()```, in his code, it will be useful to have this. I was discussing the same with @lukpueh yesterday. Hence, I added it.

SantiagoTorres replied:
Ok, makes sense.

isachit commented on inspect_return_value.py:
```diff
@@ -0,0 +1,172 @@
+      return integer <= imported_link.return_value
+
+    else:
+        raise Exception(
```
Hence, I didn't add it in the code before, but just in the latest 3-4 commits.

isachit committed:
Removed usage setting of parser, used metavar instead

SantiagoTorres commented on inspect_byproducts.py:
```diff
@@ -0,0 +1,160 @@
+  of a link file, a random string and an operator, which is used to compare
+  a certain property of the link file with the random string.
+
+  Suppose the link file is located at /user/abc/def/package.45gh325.link
```
Ha, minor nit. This is not a hex number: `45gh325`. Could you change it to, say `45fe325`? 😆

isachit replied:
Haha! Done. 👍

isachit committed:
Fixed documentation

SantiagoTorres commented:
Awesome! Thanks! 💯

isachit commented:
I don't know but it feels really good :D :D
